net/http/src/THttpLongPollEngine.cxx

```diff
@@ -0,0 +1,131 @@
+ * Copyright (C) 1995-2013, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#include "THttpLongPollEngine.h"
+
+#include "THttpCallArg.h"
+#include <TSystem.h>
+
+#include <string.h>
+
+//////////////////////////////////////////////////////////////////////////
+//                                                                      //
+// THttpLongPollEngine                                                  //
+//                                                                      //
+// Emulation of websocket with long poll requests                       //
+// Allows to send data from server to client without explicit request   //
+//                                                                      //
+//////////////////////////////////////////////////////////////////////////
+
+const char *THttpLongPollEngine::gLongPollNope = "<<nope>>";
+
+//////////////////////////////////////////////////////////////////////////
+/// returns ID of the engine, created from this pointer
+
+UInt_t THttpLongPollEngine::GetId() const
+{
+   const void *ptr = (const void *)this;
+   return TString::Hash((void *)&ptr, sizeof(void *));
+}
+
+//////////////////////////////////////////////////////////////////////////
+/// clear request, waiting for next portion of data
+
+void THttpLongPollEngine::ClearHandle()
+{
+   if (fPoll) {
+      fPoll->Set404();
+      fPoll->NotifyCondition();
+      fPoll = nullptr;
+   }
+}
+
+//////////////////////////////////////////////////////////////////////////
+/// Send binary data via connection - not supported
+
+void THttpLongPollEngine::Send(const void * /*buf*/, int /*len*/)
+{
+   Error("Send", "Binary send is not supported, use only text");
+}
+
+//////////////////////////////////////////////////////////////////////////
+/// Send const char data
+/// Either do it immediately or keep in internal buffer
+
+void THttpLongPollEngine::SendCharStar(const char *buf)
+{
+   if (fPoll) {
+      fPoll->SetContentType("text/plain");
+      fPoll->SetContent(buf);
+      fPoll->NotifyCondition();
+      fPoll = nullptr;
+   } else {
+      fBuf.push_back(std::string(buf));
+      if (fBuf.size() > 100)
+         Error("SendCharStar", "Too many send operations %d, check algorithms", (int)fBuf.size());
+   }
+}
+
+//////////////////////////////////////////////////////////////////////////////
+/// Preview data for given socket
+/// function called in the user code before processing correspondent websocket data
+/// returns kTRUE when user should ignore such http request - it is for internal use
+
+Bool_t THttpLongPollEngine::PreviewData(THttpCallArg *arg)
```

Comment:
If `arg` must be `!= nullptr` then the signature should be `THttpCallArg &arg`.

Reply:
I can use reference, but this will change interface.
Base THttpWSEngine class was introduced ~1.5 years ago.
Should we keep backward compatibility?
I have no idea if somebody else uses this handler - before it was the only way to work 
with websockets in the THttpServer.

With current design and new THttpWSHandler class user not need to access these classes at all. Should I change API at this place and hide this class from the users?

Reply:
Yes, please change this and put a note in the ReleaseNotes.